StreamChatCoreTests/Unit Tests/ClientLoggerTests.swift

```diff
@@ -0,0 +1,168 @@
+        
+        allEndpoints = [.guestToken(testUser),
+                        .addDevice(deviceId: "test", testUser),
+                        .devices(testUser),
+                        .removeDevice(deviceId: "test", testUser),
+                        .channels(.init(filter: testFilter)),
+                        .message("test"),
+                        .markAllRead,
+                        .search(.init(filter: testFilter, query: "search", pagination: .none)),
+                        .channel(.init(channel: testChannel, members: testMembers, pagination: .none, options: .all)),
+                        .stopWatching(testChannel),
+                        .updateChannel(.init(data: .init(testChannel))),
+                        .deleteChannel(testChannel),
+                        .hideChannel(testChannel, testUser, true),
+                        .showChannel(testChannel, testUser),
+                        .sendMessage(testMessage, testChannel),
+                        .sendImage("test", "image", testData, testChannel),
+                        .sendFile("test", "image", testData, testChannel),
+                        .deleteImage(testUrl, testChannel),
+                        .deleteFile(testUrl, testChannel),
+                        .markRead(testChannel),
+                        .sendEvent(.channelDeleted, testChannel),
+                        .sendMessageAction(.init(channel: testChannel, message: testMessage, action: .init(name: "test", value: "test", style: .default, type: .button, text: "test"))),
+                        .addMembers(testMembers, testChannel),
+                        .removeMembers(testMembers, testChannel),
+                        .invite(testMembers, testChannel),
+                        .inviteAnswer(.init(channel: testChannel, accept: nil, reject: nil, message: testMessage)),
+                        .replies(testMessage, .none),
+                        .deleteMessage(testMessage),
+                        .addReaction(.angry, testMessage),
+                        .deleteReaction(.angry, testMessage),
+                        .flagMessage(testMessage),
+                        .unflagMessage(testMessage),
+                        .users(.init(filter: testFilter)),
+                        .updateUsers([testUser]),
+                        .muteUser(testUser),
+                        .unmuteUser(testUser),
+                        .flagUser(testUser),
+                        .unflagUser(testUser),
+                        .ban(.init(user: testUser, channel: testChannel, timeoutInMinutes: 0, reason: nil))]
+    }
+    
+    func testLogEndpointStrings() {
+        for endpoint in allEndpoints {
+            logger.log("Endpoint: \(String(describing: endpoint))", level: .debug)
+        }
+        
+        XCTAssertEqual(logged.count, allEndpoints.count)
+    }
+    
+    func testLogURLRequest() {
```

Comment:
I look at this code and think: "so what is this test doing exactly"?

- give this test a good name
- dynamic tests are convenient only for the person that writes them; when they break they are a pain unless the error message (of even better the test name) includes enough information (ie. testLogURLRequest failed on handler  muteUser is useful)
- shotgun tests are usually bad 
- this test in general is not really useful either; we are testing that `endpoint.body` can be encoded by `JSONEncoder`; why aren't we testing `[the function used to encode endpoint.body]` with bad data and make sure it does not blow up?

Reply:
Every test case is named after the logging function they're testing. This one's testing the logging function `public func log(_ request: URLRequest)` so it's named as such. Suggestions are welcome!

Can you clarify "dynamic tests"? Do you mean looping over all possible endpoints?

Can you clarify on "shotgun tests"? It's a term I haven't heard before and could not find by searching.

The part where we create the request is taken directly from `Client+Request.swift:194`, `encodeRequest` function. During usual use, `log(_ request: URLRequest)` is used in conjuction with `encodeRequest` so I wanted the test to mirror the real world use case. So it's not testing `JSONEncoder` (the test request is fairly simple and encoding won't fail). We can also get rid of that part and just test logging without body, that'd also work for our case, although won't mirror real world use this closely. What do you think?